docs/usage/topology_aware_routing.md

```diff
@@ -0,0 +1,96 @@
+- The algorithm in the EndpointSlice controller is based on a CPU-balance heuristic. From the TopologyAwareHints documentation:
+   > The controller allocates a proportional amount of endpoints to each zone. This proportion is based on the allocatable CPU cores for nodes running in that zone. For example, if one zone had 2 CPU cores and another zone only had 1 CPU core, the controller would allocate twice as many endpoints to the zone with 2 CPU cores.
+
+   In case it is not possible to achieve a balanced distribution of the endpoints, as a safeguard mechanism the controller removes hints from the EndpointSlice resource.
+   In our setup the clients and the servers are well-known and usually the traffic a component receives does not depend on the zone's allocatable CPU. 
+   Many components deployed by Gardener are scaled automatically by VPA. In case of an overload of a replica, the VPA should provide and apply enhanced CPU and memory resources. Additionally, Gardener uses the cluster-autoscaler to upscale/downscale Nodes dynamically. Hence, it is not possible to ensure a balanced allocatable CPU across the zones.
+- TopologyAwareHints does not work at low-endpoint counts.
+- Hints provided by the EndpointSlice controller are not deterministic. With cluster-autoscaler running and load increasing, hints can be removed in the next moment. There is no option to enforce the zone-level topology.
+
+For more details, see the following issue [kubernetes/kubernetes#113731](https://github.com/kubernetes/kubernetes/issues/113731).
+
+To circumvent these issues with the EndpointSlice controller, a mutating webhook in the gardener-resource-manager assigns hints to EndpointSlice resources. For each endpoint in the EndpointSlice it sets the endpoint's hints to the endpoint's zone. The webhook overwrites the hints provided by the EndpointSlice controller in kube-controller-manager. For more details, see the [webhook's documentation](../concepts/resource-manager.md#endpointslice-hints).
+
+##### kube-proxy
+
+By default, with kube-proxy running in `iptables` mode, traffic is distributed randomly across all endpoints, regardless of where it originates from. In a cluster with 3 zones, traffic is more likely to go to another zone than to stay in the current zone.
+With the `TopologyAwareRouting` feature, kube-proxy filters the endpoints it routes to based on the hints in the EndpointSlice resource. In most of the cases, kube-proxy will prefer the endpoint(s) in the same zone. For more details, see the [following section](https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/#implementation-kube-proxy).
+
+## Topology-aware Services in the Seed cluster
```

Comment:
What do you think about handling hints for every service in a multi-zone control plane? For some it wouldn't be relevant but at least it makes the entire handling easier and we wouldn't miss a service.

Reply:
In the current revision of the PR, hints will be added to only if the feature gate is enabled and the Shoot is HA with failure tolerance type `zone`. A question that is related: I was also not sure how the extensions will make a Service topology-aware. The provider extensions can make the `csi-snapshot-validation` Service topology-aware. I was not sure how to handle this. One option is to add the label and annotation directly - this would mean that the Service could have hints when the gardenlet feature is disabled. Other option is to try to duplicate the "gate" logic in the extensions - the "gate" logic is the feature gate is enabled and the control plane is HA with type `zone`. This options looks like an overkill.
I am mentioning all of this because if we decide to make each Service in Shoot control plane the topology-aware, then we would also tackle this problem with how to make Services topology-aware outside of gardenlet.

---

With regards to the "feature gate" - I put the feature behind a feature gate because we wanted to have the control when the enable/disable the functionality. But in the longterm I think we will want to preserve this behaviour and I rather see that the feature gate can be transformed to a config option (similar to what happened with the Logging feature gate).

---

With regards to your question - I assume it should work. It should be able to fallback to the default behaviour when there is no endpoint in the given zone.
So what is the suggestion? Another webhook in GRM that makes all Services for HA Shoot with `zone` topology-aware?

Reply:
> So what is the suggestion? Another webhook in GRM that makes all Services for HA Shoot with zone topology-aware?

Isn't this the easier option? We were wondering why we are differentiating here and having custom code for a few services only. For `zone` shoots, wouldn't it simplify things if we make all services topology-aware? If you don't see issues/risks with this, then let's probably go for it. It's still all hidden behind the feature gate, so if it unexpectedly doesn't work we can change the approach?

Reply:
> Isn't this the easier option? We were wondering why we are differentiating here and having custom code for a few services only.

I started initially with this approach because in the description of https://github.com/gardener/gardener/issues/6718 as substasks were listed the Services that need to adapted - `etcd-{main,events}-client`, `kube-apiserver`, `gardener-resource-manager`, `vpa-webhook`.
I don't know which is the easier option. If we decide on the webhook approach, how would we handle the garden cluster? Do I need to still preserve the existing logic in the kube-apiserver, etcd and GRM webhooks as I assume the gardener-operator is reusing these components?
Additionally, there is a case where components are running outside of the Shoot control plane in the Seed. In the early days I sketched that the provider-extension Pods in the Seed can be also topology-aware:
```
##### Extensions with a Shoot webhook (gardener-extension-provider-aws, gardener-extension-provider-alicloud, gardener-extension-provider-vsphere, gardener-extension-networking-cilium)

Some extensions like gardener-extension-provider-aws and gardener-extension-provider-alicloud serve Shoot webhooks. The provider-aws Shoot webhook mutates the `addons-nginx-ingress-controller` ConfigMap. provider-alicloud mutates the `vpn-shoot` and `addons-nginx-ingress-controller` Services and the `metrics-server` Deployment. networking-cilium mutates the `node-local-dns` DaemonSet and ConfigMap. provider-vsphere mutates the `metrics-server` Deployment.
```

Which are the other Services in HA control plane type zone that can benefit from such thing? I see Services for controllers (`cert-controller-manager`, `cloud-controller-manager`, `cluster-autoscaler`, `kube-controller-manager`, `kube-scheduler`, `machine-controller-manager`) where the controllers run 2 replicas. I think only prometheus is calling these Services. I don't know how prometheus is working internally - whether it calls the Service or whether it directly fetching metrics from the Pods. I can try to clarify with the monitoring colleagues whether prometheus could benefit from controller Services being topology-aware.
Pods backing Services `grafana-operators`, `grafana-users`, `kube-state-metrics`, `loki`, `prometheus-web`, `vpn-seed-server-0`, `vpn-seed-server-1` run with 1 replica, hence they cannot benefit from topology-aware routing. The single running replica will be always called.
With this PR we already tackle (`kube-apiserver`, `etcd-{main,events}-client`, `gardener-resource-manager`, `vpa-webhook`) and we have plans to tackle `csi-snapshot-validation` in the extensions.
I don't know who is consuming Serivces `etcd-{main,events}-peer` and whether they can benefit from being topology aware. I can try to clarify with the etcd colleagues whether this peer Service can benefit from topology aware routing.

```
NAME                         TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                       AGE
cert-controller-manager      ClusterIP   None             <none>        10258/TCP                     91d
cloud-controller-manager     ClusterIP   None             <none>        10258/TCP                     91d
cluster-autoscaler           ClusterIP   None             <none>        8085/TCP                      91d
csi-snapshot-validation      ClusterIP   10.243.219.255   <none>        443/TCP                       91d
etcd-events-client           ClusterIP   10.243.224.57    <none>        2379/TCP,2380/TCP,8080/TCP    91d
etcd-events-peer             ClusterIP   None             <none>        2380/TCP                      91d
etcd-main-client             ClusterIP   10.243.106.101   <none>        2379/TCP,2380/TCP,8080/TCP    91d
etcd-main-peer               ClusterIP   None             <none>        2380/TCP                      91d
gardener-resource-manager    ClusterIP   10.243.142.119   <none>        8080/TCP,8081/TCP,443/TCP     91d
grafana-operators            ClusterIP   10.243.171.181   <none>        3000/TCP                      91d
grafana-users                ClusterIP   10.243.175.244   <none>        3000/TCP                      91d
kube-apiserver               ClusterIP   10.243.60.15     <none>        443/TCP                       91d
kube-controller-manager      ClusterIP   None             <none>        10257/TCP                     91d
kube-scheduler               ClusterIP   10.243.71.38     <none>        10259/TCP                     91d
kube-state-metrics           ClusterIP   10.243.165.190   <none>        80/TCP                        91d
loki                         ClusterIP   10.243.247.97    <none>        3100/TCP,8080/TCP,9273/TCP    91d
machine-controller-manager   ClusterIP   None             <none>        10258/TCP                     91d
prometheus-web               ClusterIP   10.243.7.28      <none>        80/TCP                        91d
vpa-webhook                  ClusterIP   10.243.7.166     <none>        443/TCP                       91d
vpn-seed-server-0            ClusterIP   10.243.62.162    <none>        1194/TCP,9443/TCP,15000/TCP   51d
vpn-seed-server-1            ClusterIP   10.243.55.44     <none>        1194/TCP,9443/TCP,15000/TCP   51d
```

Feel free to comment if a missed a Service that can be topology-aware.

In summary if `kube-apiserver`, `etcd-{main,events}-client`, `gardener-resource-manager`, `vpa-webhook` are the only Services that we need to be topology-aware and if I have to preserve the logic in the components for the gardener-operator, do we benefit from the webhook approach?
For webhooks, we also had a discussion in the past. IMO in some cases it is cheaper to do the changes in the components/charts explicitly than introducing yet another webhook that does the mutation over the network.

PS. I am just trying to figure out which is the better approach to prevent going back and forth.

Reply:
From my perspective, the webhook approach has the clear advantage that new services are directly covered. Hence, we do not need to keep yet another topic in mind when extending gardener.
On the downside, the webhook approach increases the load on `kupe-apiserver` due to additional calls and also to a small extent `etcd` due to the added hints. Personally, I do not think that these downsides are stronger argument than the reduction in cognitive load caused by yet another topic to keep in mind when creating new services.
WDYT?

Reply:
In https://github.com/gardener/gardener/pull/7191#discussion_r1094546811 I shared the following concerns that are not answered in your reply:
- how to tackle the garden cluster with the webhook approach?
- how to tackle Seed components running outside of the Shoot control plane with the webhook approach?

Reply:
We had a sync last week and agreed on the consensus to put the TopologyAwareRouting functionality behind a Seed setting. The motivation for a Seed setting comes from the fact that this approach will allow extensions to make their components topology-aware (for example by reading the Seed spec from the Cluster resource and checking if topologyAwareRouting is enabled for the Seed).
The webhook approach has the benefit that there is no need to cover each and every Service. But it has the following uncertainties: it can potentially cause issue if a Service does not need to be topology-aware; it is not optimal performance-wise (the mutation happens over the network; yet another failure point).